### PR TITLE
Bind service

### PR DESCRIPTION
**Pull request by MHBauer**
creating so we can watch along easier

 - [ ] document your feature
 - [ ] comment your code
 - [ ] update the README / how-to-run
 - [ ] have tests
 - [ ] get reviews

### CODE REVIEW

File: pkg/containermanager/docker/docker_manager.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ func NewDockerContainerManager(logger lager.Logger, client DockerClient) contain
 	}
 }
 
-func (dc dockerContainerManager) Provision(ctx context.Context, containerConfig *containermanager.ContainerConfig) error {
+func (dc dockerContainerManager) Provision(ctx context.Context, containerConfig containermanager.ContainerConfig) error {
```
Comment (MHBauer) [inline pkg/containermanager/docker/docker_manager.go]: is there a reason for 'not a pointer' we can pull out into a separate pr to reduce review churn?

Comment (nimakaviani) [reply pkg/containermanager/docker/docker_manager.go]: yeah I made that part of a separate commit to maybe later on split into a separate PR. generally I think configs should be treated as immutable objects.

I had it as a meeting topic for us to discuss since I have noticed that I usually make objects not be pointers unless needed, and that they get changed somehow during the process. Would be good to establish a team convention to avoid drift.

Comment (MHBauer) [reply pkg/containermanager/docker/docker_manager.go]: I don't think there's a good go way to do that exact thing. The closet thing I can think of is to make a config interface with accessors.

pointer has the concept of you can 'accidentally' set something in the passed in struct, and copy makes a copy and can be confusing about why you changed something and it did not change.

File: pkg/containermanager/docker/docker_manager.go [hunk 2]
```diff
@@ -94,6 +98,59 @@ func (dc dockerContainerManager) Deprovision(ctx context.Context, instanceID str
...
+			}
+		}
+	}
+
+	file, err := ioutil.TempFile("", uuid.New())
+	if err != nil {
+		return nil, err
+	}
+	downloadFile(file.Name(), bindingConfig.Params.ContractUrl)
+	defer os.RemoveAll(file.Name())
```
Comment (MHBauer) [inline pkg/containermanager/docker/docker_manager.go]: this and the previous five lines or so should move into the generic part of bind. It's not docker specific and needs to be reused by any other containermanagers.

File: pkg/deployer/deployer.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+func NewEthereumDeployer(logger lager.Logger, config config.Config) Deployer {
+	return &ethereumDeployer{
+		logger: logger,
+		config: config,
+	}
+}
+
+func (e ethereumDeployer) DeployContract(contractInfo *ContractInfo, containerInfo *containermanager.ContainerInfo) (*NodeInfo, error) {
```
Comment (MHBauer) [inline pkg/deployer/deployer.go]: bunch of files being accessed in here to do  I-don't-know-what.

File: pkg/broker/broker.go [hunk 4]
```diff
@@ -102,7 +116,59 @@ func (b BlockheadBroker) Deprovision(ctx context.Context, instanceID string, det
...
+
+	service := b.state.Services[details.ServiceID]
+	if service == nil {
+		return brokerapi.Binding{}, errors.New("service not found")
+	}
+
+	plan := service.Plans[details.PlanID]
+	if plan == nil {
+		return brokerapi.Binding{}, errors.New("plan not found")
+	}
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/broker/broker.go]: checking now will at least make things easier when we stop using the library

File: pkg/deployer/deployer.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		Provider string   `json:"provider"`
+		Password string   `json:"password"`
+		Args     []string `json:"args"`
+	}{
+		Provider: fmt.Sprintf("http://%s:%s", containerInfo.IP, "8545"),
+		Password: "",
+		Args:     contractInfo.ContractArgs,
+	}
+
+	configFile, err := ioutil.TempFile("", uuid.New())
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/deployer/deployer.go]: i think its the config file for pusher.js. was there a reason we switched to a config file versus just passing in arguments?

File: cmd/broker/main.go [hunk 6]
```diff
@@ -36,9 +37,10 @@ func main() {
 		logger.Fatal("docker-client", err)
 	}
 
-	manager := docker.NewDockerContainerManager(logger, cli)
+	manager := docker.NewDockerContainerManager(logger, cli, state.Config.DeployerPath)
+	deployer := deployer.NewEthereumDeployer(logger, state.Config)
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply cmd/broker/main.go]: It is good to have the deployer interface if it happens that we want to bring in other blockchains.

File: pkg/broker/broker.go [hunk 7]
```diff
@@ -102,7 +116,59 @@ func (b BlockheadBroker) Deprovision(ctx context.Context, instanceID string, det
...
+	err := json.Unmarshal(details.RawParameters, contractInfo)
+	if err != nil {
+		return brokerapi.Binding{}, err
+	}
+
+	if contractInfo.ContractUrl == "" {
+		return brokerapi.Binding{}, errors.New("contract_url not found")
+	}
+
+	file, err := ioutil.TempFile("", uuid.New())
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply pkg/broker/broker.go]: read the tests

File: cmd/broker/main_suite_test.go [hunk 8]
```diff
@@ -90,9 +90,10 @@ var _ = SynchronizedBeforeSuite(func() []byte {
```
Comment (swetharepakula) [inline cmd/broker/main_suite_test.go]: Can we add something in the suite test to make sure the right node packages are installed? Or add that to the makefile or something?

File: pkg/containermanager/container_manager.go [hunk 9]
```diff
@@ -10,8 +10,24 @@ type ContainerConfig struct {
...
+type ContainerInfo struct {
+	IP       string
+	Bindings map[string][]Binding
+}
+
 //go:generate counterfeiter -o ../fakes/fake_container_manager.go . ContainerManager
 type ContainerManager interface {
 	Provision(ctx context.Context, cc ContainerConfig) error
 	Deprovision(ctx context.Context, instanceID string) error
+	Bind(cts context.Context, bc BindConfig) (*ContainerInfo, error)
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply pkg/containermanager/container_manager.go]: It isn't. +1 on changing the name

File: pkg/containermanager/docker/docker_manager.go [hunk 10]
```diff
@@ -94,6 +96,32 @@ func (dc dockerContainerManager) Deprovision(ctx context.Context, instanceID str
 	return nil
 }
 
+func (dc dockerContainerManager) Bind(ctx context.Context, bindingConfig containermanager.BindConfig) (*containermanager.ContainerInfo, error) {
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply pkg/containermanager/docker/docker_manager.go]: we should take existing network config (port / host).

What we have works if the container engine uses bridge network. Not sure if it is generic enough for other cases. Need to check!

File: pkg/deployer/deployer.go [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	ContractUrl  string   `json:"contract_url"`
+	ContractArgs []string `json:"contract_args"`
+	ContractPath string
+}
+
+type NodeInfo struct {
+	Account          string `json:"address"`
+	Interface        string `json:"abi"`
+	ConctractAddress string `json:"contract_address"`
+	GasPrice         string `json:"gas_price"`
```
[the remark this answers is not in the document]
Comment (nimakaviani) [reply pkg/deployer/deployer.go]: The gas price for deploying the contract. We had it returned in the previous broker so I included it here for us to decide what to do about it. We may not need it.

File: pkg/deployer/deployer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (nimakaviani) [reply pkg/deployer/deployer.go]: here is the list

- the temp file to download the contract to
- a temp file for pusher.js config (it makes it easier not to have to deal with spaces in the passed in arguments. something we had problems with in the old ruby code)
- a temp file for output of pusher.js to be written to as json and the deployer to read off of it (better than having to read the output text and parsing it into a json object)

File: cmd/broker/main_suite_test.go (hunk 8 above)
Comment (nimakaviani) [reply cmd/broker/main_suite_test.go]: thats a good idea. will look into it.

File: pkg/deployer/deployer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/deployer/deployer.go]: Okay, just wasn't clear what the value was for. I don't think we have to remove it. But probably don't need it until we connect to a public network.

File: pkg/deployer/deployer.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/deployer/deployer.go]: I think that sounds reasonable to me

File: cmd/broker/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (swetharepakula) [reply cmd/broker/main.go]: i think thats a good idea. we aren't getting vendor locked ourselves.

File: images/Dockerfile.test [hunk 12]
```diff
@@ -13,4 +13,6 @@ RUN npm install --production --quiet solc@0.4.25
 workdir /go/src/github.com/cloudfoundry-incubator/blockhead
 copy . .
 
-cmd ["go", "test", "-v", "./..."]
+RUN go install ./vendor/github.com/onsi/ginkgo/ginkgo
+
+cmd ["ginkgo", "-r", "-randomizeAllSpecs", "--randomizeSuites", "--keepGoing", "--cover", "--trace", "--progress"]
```
Comment (MHBauer) [inline images/Dockerfile.test]: :ok_hand:

File: cmd/broker/main_suite_test.go [hunk 13]
```diff
@@ -88,7 +88,7 @@ var _ = SynchronizedBeforeSuite(func() []byte {
 	for lib, version := range nodeLibVersionMap {
 		cmd := exec.Command("npm", "list", lib)
 		output, err := cmd.CombinedOutput()
-		Expect(err).NotTo(HaveOccurred())
+		Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Errored when verifying npm modeule: %s", lib))
```
Comment (MHBauer) [inline cmd/broker/main_suite_test.go]: :ok_hand:
